istio.io testing overhaul (#18015)
istio/istio: tests/integration/istioio/README.md

be run as part of the resulting test function:

```golang
func TestCombinedMethods(t *testing.T) {
    framework.
        NewTest(t).
        Run(istioio.NewBuilder().
            // Run a script and create a snippet.
            Add(istioio.Command{
                Input:         istioio.Path("myscript.sh"),
                WorkDir:       env.IstioSrc,
                CreateSnippet: true,
            }, 
            // Wait for all pods in namespace foo to start.
            istioio.MultiPodWait("foo"),
            // Run another script and verify the results.
            istioio.Command{
                Input:         istioio.Path("myotherscript.sh"),
                WorkDir:       env.IstioSrc,
                CreateSnippet: true,
                Verify:        istioio.TokenVerifier(istioio.Path("myotherscript_verify.txt`),
            }).Build())
}
```

## Builder

The `istioio.NewBuilder` returns a `istioio.Builder` that is used to build an Istio
test run function and has the following methods:

- `Add`: adds a step to the test.
- `Defer`: provides a step to be run after the test completes.
- `Build`: builds an Istio test run function.

## Selecting Input

> Many of the test steps require an `Input` which they obtain from an

Reviewer: ```suggestion
Many test steps require an `Input` which they obtain from an
```

`istioio.InputSelector`:

```golang
type Input interface {
	InputSelector
	Name() string
	ReadAll() (string, error)
}

type InputSelector interface {
	SelectInput(Context) Input
}
```

Some common `InputSelector` implementations include:

- `istioio.Inline`: allows you to inline the content for the `Input` directly in the code.
- `istioio.Path`: reads in a file from the specified path.
- `istioio.BookInfo`: is like `istioio.Path` except that the value is assumed to be
relative to the BookInfo source directory (`$GOPATH/src/istio.io/istio/samples/bookinfo/platform/kube/`).

> `InputSelector` is provided a `istioio.Context` at runtime, which it can use to

Reviewer: ```suggestion
An `InputSelector` provides an `istioio.Context` at runtime, which can be used to
```
